cmd/diff.go

```diff
@@ -0,0 +1,174 @@
+*/
+
+package cmd
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"reanahub/reana-client-go/client"
+	"reanahub/reana-client-go/client/operations"
+	"reanahub/reana-client-go/utils"
+
+	"github.com/jedib0t/go-pretty/v6/text"
+
+	"github.com/spf13/cobra"
+)
+
+const diffDesc = `
+Show diff between two workflows.
+
+The ` + "``diff``" + ` command allows to compare two workflows, the workflow_a and
+workflow_b, which must be provided as arguments. The output will show the
+difference in workflow run parameters, the generated files, the logs, etc.
+
+Examples:
+
+	$ reana-client diff myanalysis.42 myotheranalysis.43
+
+	$ reana-client diff myanalysis.42 myotheranalysis.43 --brief
+`
+
+type diffOptions struct {
+	token     string
+	workflowA string
+	workflowB string
+	brief     bool
+	unified   int
+}
+
+// newDiffCmd creates a command to show diff between two workflows.
+func newDiffCmd() *cobra.Command {
+	o := &diffOptions{}
+
+	cmd := &cobra.Command{
+		Use:   "diff",
+		Short: "Show diff between two workflows.",
+		Long:  diffDesc,
+		Args:  cobra.ExactArgs(2),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			o.workflowA = args[0]
+			o.workflowB = args[1]
+			return o.run(cmd)
+		},
+	}
+
+	f := cmd.Flags()
+	f.StringVarP(&o.token, "access-token", "t", "", "Access token of the current user.")
+	f.BoolVarP(&o.brief, "brief", "q", false, `If not set, differences in the contents of the
+files in the two workspaces are shown.`)
+	f.IntVarP(
+		&o.unified, "unified", "u", 5, "Sets number of context lines for workspace diff output.",
+	)
+
+	return cmd
+}
+
+func (o *diffOptions) run(cmd *cobra.Command) error {
+	diffParams := operations.NewGetWorkflowDiffParams()
+	diffParams.SetAccessToken(&o.token)
+	diffParams.SetWorkflowIDOrNamea(o.workflowA)
+	diffParams.SetWorkflowIDOrNameb(o.workflowB)
+	diffParams.SetBrief(&o.brief)
+	contextLines := fmt.Sprintf("%d", o.unified)
+	diffParams.SetContextLines(&contextLines)
+
+	api, err := client.ApiClient()
+	if err != nil {
+		return err
+	}
+	diffResp, err := api.Operations.GetWorkflowDiff(diffParams)
+	if err != nil {
+		return err
+	}
+
+	err = displayDiffPayload(cmd, diffResp.Payload)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func displayDiffPayload(cmd *cobra.Command, p *operations.GetWorkflowDiffOKBody) error {
+	leadingMark := "==>"
+
+	if p.ReanaSpecification != "" {
+		var specificationDiff map[string][]string
+		err := json.Unmarshal([]byte(p.ReanaSpecification), &specificationDiff)
+		if err != nil {
+			return err
+		}
+
+		// Rename section workflow to specification
+		val, hasWorkflow := specificationDiff["workflow"]
+		if hasWorkflow {
+			specificationDiff["specification"] = val
+			delete(specificationDiff, "workflow")
+		}
+		equalSpecification := true
+		for section, lines := range specificationDiff {
+			if len(lines) != 0 {
+				equalSpecification = false
+				utils.PrintColorable(
+					fmt.Sprintf("%s Differences in workflow %s\n", leadingMark, section),
```

Comment:
One thing I noticed with respect to the Python client is the difference section ordering:

```console
$ reana-client diff events.1 test | grep "^==>"
==> Differences in workflow inputs
==> Differences in workflow outputs
==> Differences in workflow version
==> Differences in workflow workspace
==> Differences in workflow specification
==> Differences in workflow workspace

$ $ ./reana-client-go diff events.1 test | grep "==>"
==> Differences in workflow specification
==> Differences in workflow inputs
==> Differences in workflow outputs
==> Differences in workflow version
==> Differences in workflow workspace
==> Differences in workflow workspace
```

Some observations:

- It's weird that there are two "workflow workspace" sections, but that's also true for the Python client, and should be fixed there. So nothing to do here.
- Can we align the output order to respect Python client?

Reply:
The ordering is different because GO's maps are unordered, unlike in Python. It's possible to maintain order by using [json.Decoder](https://pkg.go.dev/encoding/json#Decoder) instead of the [json.Unmarshal](https://pkg.go.dev/encoding/json#Unmarshal) being used. The code will probably be a little bit more verbose but the change shouldn't be too difficult

Reply:
Shall we do it now or after package refactoring?

Reply:
We can do it now

Reply:
Turns out I was wrong about using the Decoder, it works nicely for arrays but not so much for objects, since they don't guarantee order by nature. By looking at [this discussion](https://github.com/golang/go/issues/27179) I found some hacky workarounds but no nice solution for this, the developers don't seem to support it because for order one should use an array.

An example of not so nice solution, is using the Decoder and a manually parse of the JSON, it doesn't look good at all: https://go.dev/play/p/jOXpSDpJz04

Reply:
I'm not sure if it's worth implementing this order (if we want order, shouldn't the server return it as an array?)

Reply:
Yeah, fully agreed that it would be very clean to do this on the server-side. But since it would necessitate breaking REST API response changes, it's not something for now.

Hmm, looking at the pasted workaround, it might do actually as a stop-gap solution; however it's also possible that we can live with the Python client <-> Go client output differences for this commend, since it is not probable that some clients has wrapped it into some wider scripted glue usage scenarios... I'm kind of 60:40 split on this.